servant/src/Servant/API/Stream.hs

```diff
@@ -70,143 +70,173 @@ data Stream (method :: k1) (status :: Nat) (framing :: *) (contentType :: *) (a
+-- or errors that terminates with a Nothing.
+type SourceIO = SourceT IO
 
--- | ToStreamGenerator is intended to be implemented for types such as Conduit, Pipe, etc. By implementing this class, all such streaming abstractions can be used directly as endpoints.
-class ToStreamGenerator a b | a -> b where
-  toStreamGenerator :: a -> StreamGenerator b
+-- | ToSourceIO is intended to be implemented for types such as Conduit, Pipe, etc. By implementing this class, all such streaming abstractions can be used directly as endpoints.
+class ToSourceIO chunk a | a -> chunk where
+    toSourceIO :: a -> SourceIO chunk
 
-instance ToStreamGenerator (StreamGenerator a) a where
-  toStreamGenerator x = x
+-- | Auxiliary class for @'ToSourceIO' x ('SourceT' m x)@ instance.
+class SourceToSourceIO m where
+    sourceToSourceIO :: SourceT m a -> SourceT IO a
 
-instance ToStreamGenerator (NonEmpty a) a where
-  toStreamGenerator (x :| xs) = StreamGenerator $ \f g -> f x >> traverse_ g xs
+instance SourceToSourceIO IO where
+    sourceToSourceIO = id
 
-instance ToStreamGenerator [a] a where
-  toStreamGenerator []       = StreamGenerator $ \_ _ -> return ()
-  toStreamGenerator (x : xs) = StreamGenerator $ \f g -> f x >> traverse_ g xs
+-- | Relax to use auxiliary class, have m
+instance SourceToSourceIO m => ToSourceIO chunk (SourceT m chunk) where
+    toSourceIO = sourceToSourceIO
 
--- | Clients reading from streaming endpoints can be implemented as producing a
--- @ResultStream@ that captures the setup, takedown, and incremental logic for
--- a read, being an IO continuation that takes a producer of Just either values
--- or errors that terminates with a Nothing.
-newtype ResultStream a = ResultStream { runResultStream :: forall b. (IO (Maybe (Either String a)) -> IO b) -> IO b }
-
--- | FromResultStream is intended to be implemented for types such as Conduit, Pipe, etc. By implementing this class, all such streaming abstractions can be used directly on the client side for talking to streaming endpoints.
-class FromResultStream a b | b -> a where
-  fromResultStream :: ResultStream a -> IO b
-
-instance FromResultStream a (ResultStream a) where
-  fromResultStream = return
-
--- | Uses 'unsafeInterleaveIO'
-instance FromResultStream a [a] where
-  fromResultStream x = runResultStream x lazyRead
-
--- | Uses 'unsafeInterleaveIO'
-instance FromResultStream a (NonEmpty a) where
-  fromResultStream x = runResultStream x $ \r -> do
-      e <- r
-      case e of
-          Nothing        -> fail "Empty stream"
-          Just (Left er) -> fail er
-          Just (Right y) -> do
-              ys <- lazyRead r
-              return (y :| ys)
-
-lazyRead :: IO (Maybe (Either String a)) -> IO [a]
-lazyRead r = go
-  where
-    go = unsafeInterleaveIO loop
+instance ToSourceIO a (NonEmpty a) where
+    toSourceIO (x :| xs) = fromStepT (Yield x (foldr Yield Stop xs))
+
+instance ToSourceIO a [a] where
+    toSourceIO = source
+
```

Comment:
I suppose we'll want to support all the reasonable container types we directly or indirectly depend on?

Reply:
I guess things in `base` would be enough. But I still wonder how #841 would reshuffle this.

Let's leave an issue to remember about other containers?

Reply:
OK.